plugins/genesys/fu-genesys-usbhub-dev-firmware.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2023 Adam.Chen <Adam.Chen@genesyslogic.com.tw>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-genesys-common.h"
+#include "fu-genesys-usbhub-dev-firmware.h"
+#include "fu-genesys-usbhub-firmware.h"
+#include "fu-genesys-usbhub-struct.h"
+
+struct _FuGenesysUsbhubDevFirmware {
+	FuFirmwareClass parent_instance;
+};
+
+G_DEFINE_TYPE(FuGenesysUsbhubDevFirmware, fu_genesys_usbhub_dev_firmware, FU_TYPE_FIRMWARE)
+
+static gboolean
+fu_genesys_usbhub_dev_firmware_check_magic(FuFirmware *firmware,
+					   GBytes *fw,
+					   gsize offset,
+					   GError **error)
+{
+	guint8 magic[GENESYS_USBHUB_FW_SIG_LEN] = {0x0};
+
+	if (!fu_memcpy_safe(magic,
+			    sizeof(magic),
+			    0, /* dst */
+			    g_bytes_get_data(fw, NULL),
+			    g_bytes_get_size(fw),
+			    offset + GENESYS_USBHUB_FW_SIG_OFFSET,
+			    sizeof(magic),
+			    error)) {
+		g_prefix_error(error, "failed to read magic: ");
+		return FALSE;
+	}
+	if (memcmp(magic, GENESYS_USBHUB_FW_SIG_TEXT_DEV_BRIDGE, sizeof(magic)) != 0) {
```

Comment:
can you use fu_memcmp_safe() here directly?

Reply:
Ohh no, ignore me -- you're using an offset.

Reply:
like this ?
```
	if (fu_memcmp_safe(magic,
			   sizeof(magic),
			   GENESYS_USBHUB_FW_SIG_TEXT_DEV_BRIDGE,
			   GENESYS_USBHUB_FW_SIG_LEN,
			   error) == 0) {
```

Reply:
I'll rework fu_memcmp_safe() to allow an offset. Give me a few mins.

Reply:
How about https://github.com/fwupd/fwupd/pull/5940